### DISCUSSION TITLE
should we mock webpack internal plugin

### DISCUSSION THREAD

**Discussion by hardfist** [web-infra-dev/rspack#3252] [Ideas]
lots of community use webpack's internal plugin like definePlugin, which is not supported in rspack, so we need to fork&clone these plugins, if we mock webpack's internal plugin it may make the plugin compatible much easier

**Comment by hardfist** [comment]
cc @web-infra-dev/rspack

**Comment by underfin** [comment]
I want this, it is awesome!

**Comment by ahabhgk** [comment]
```
        js side                                                         rust side
new rspack.DefinePlugin(options) -> ["builtin:define", options] -> get_builtin_plugin("builtin:define")::new(options)
```

just like what we do for builtin loader

**Reply by underfin** [reply]
I think `webpack.DefinePlugin` is better,the user doesn't change anything.

**Reply by hardfist** [reply]
or we could change options in js side?

**Reply by ahabhgk** [reply]
> or we could change options in js side?

bad idea

**Reply by ahabhgk** [reply]
Another benefit is we can expose more internal plugin in this way, EntryPlugin, SourceMapDevToolPlugin, etc. and for now `RawOptionsApply` has two responsibility: 1. js options (RawOptions) to rust options (CompilerOptions), 2. apply internal (rust) plugins, in this way `RawOptionsApply` can only have the first responsibility

**Comment by hardfist** [comment]
conclusion we all agree we should implement proxy plugin
